src/java/org/apache/cassandra/index/sai/disk/StorageAttachedIndexWriter.java

```diff
@@ -0,0 +1,255 @@
+    private boolean aborted = false;
+
+    private long sstableRowId = 0;
+
+    public StorageAttachedIndexWriter(IndexDescriptor indexDescriptor,
+                                      Collection<StorageAttachedIndex> indexes,
+                                      LifecycleNewTracker lifecycleNewTracker) throws IOException
+    {
+        this(indexDescriptor, indexes, lifecycleNewTracker, false);
+    }
+
+    public StorageAttachedIndexWriter(IndexDescriptor indexDescriptor,
+                                      Collection<StorageAttachedIndex> indexes,
+                                      LifecycleNewTracker lifecycleNewTracker,
+                                      boolean perIndexComponentsOnly) throws IOException
+    {
+        this.indexDescriptor = indexDescriptor;
+        this.primaryKeyFactory = indexDescriptor.primaryKeyFactory;
+        this.indexes = indexes;
+        this.rowMapping = RowMapping.create(lifecycleNewTracker.opType());
+        this.perIndexWriters = indexes.stream().map(index -> indexDescriptor.newPerIndexWriter(index,
+                                                                                               lifecycleNewTracker,
+                                                                                               rowMapping))
+                                      .filter(Objects::nonNull) // a null here means the column had no data to flush
+                                      .collect(Collectors.toList());
+
+        // If the SSTable components are already being built by another index build then we don't want
+        // to build them again so use a null writer
+        this.perSSTableWriter = perIndexComponentsOnly ? PerSSTableWriter.NONE : indexDescriptor.newPerSSTableWriter();
+    }
+
+    @Override
+    public void begin()
+    {
+        logger.debug(indexDescriptor.logMessage("Starting partition iteration for storage attached index flush for SSTable {}..."), indexDescriptor.descriptor);
+        stopwatch.start();
+    }
+
+    @Override
+    public void startPartition(DecoratedKey key, long position)
+    {
+        if (aborted) return;
+        
+        currentKey = key;
+    }
+
+    @Override
+    public void nextUnfilteredCluster(Unfiltered unfiltered)
+    {
+        if (aborted) return;
+
+        // Ignore range tombstones...
+        if (!unfiltered.isRow())
+            return;
+
+        try
+        {
+            addRow((Row)unfiltered);
+        }
+        catch (Throwable t)
+        {
+            logger.error(indexDescriptor.logMessage("Failed to record a row during an index build"), t);
+            abort(t, true);
+        }
+    }
+
+    @Override
+    public void partitionLevelDeletion(DeletionTime deletionTime, long position)
+    {
+        // Deletions (including partition deletions) are accounted for during reads.
+    }
+
+    @Override
+    public void staticRow(Row staticRow, long position)
+    {
+        if (aborted) return;
```

Comment:
This might look untested in the current patch, but `IndexQuerySupport` should hit it once its on-disk bits are brought back into play, right? (i.e. `STATIC_COLUMNS`)

Reply:
Correct, those tests will be enabled properly when numeric support is added.

Saying that, I thought, initially, that you were commenting on the `abort` code path. I am going to look at this to see how much of the abort path is being tested atm.